include/ofi_coll.h

```diff
@@ -104,52 +109,67 @@ struct util_coll_reduce_item {
 	enum fi_op		op;
 };
 
-struct util_coll_comp_item;
-
-typedef void (*util_coll_comp_t)(struct util_coll_mc *coll_mc,
-				 struct util_coll_comp_item *comp);
-
-struct util_coll_join_comp_data {
-	uint64_t		cid_buf[OFI_CONTEXT_ID_SIZE];
-	uint64_t		tmp_cid_buf[OFI_CONTEXT_ID_SIZE];
-};
-
-struct util_coll_comp_item {
-	struct util_coll_hdr	hdr;
-	enum util_coll_op_type	op_type;
-	void			*data;
-	util_coll_comp_t	comp_fn;
-};
-
 struct util_coll_mc {
 	struct fid_mc		mc_fid;
 	struct fid_ep		*ep;
-	struct util_av_set 	*av_set;
+	struct util_av_set	*av_set;
 	struct slist		barrier_list;
 	struct slist		deferred_list;
-	int 			my_rank;
-	uint16_t		cid;
-	uint16_t		tag_seq;
+	struct slist		pending_xfer_list;
+	int			my_rank;
+	uint16_t		cid_id;
+	uint16_t		cid_seq;
 	ofi_atomic32_t		ref;
 };
 
+struct util_coll_state;
+
+typedef void (*util_coll_comp_fn_t)(struct util_coll_state *state);
+
+struct util_coll_cid_data {
+	uint64_t	cid_buf[OFI_CONTEXT_ID_SIZE];
+	uint64_t	tmp_cid_buf[OFI_CONTEXT_ID_SIZE];
+};
```

Comment:
This should store something more useful than arbitrary arrays of u64.  What is actually being stored here?

Reply:
My understanding is that the intent is to store the equivalent of MPI's context ID mask, which is an array of (64) 32-bit integers, and will be used for join's reduction.  I think the tmp part is probably not needed, so I will remove.

there is also a context id [mask] stored locally for each rank to track available cids.  do you think some structure like the below would help clarify this? 

```
#define OFI_MAX_CONTEXT_MASK 4
struct context_id_mask {
        uint64_t values[OFI_MAX_CONTEXT_MASK];
};
```

If so, this structure would become the buffer for the join data and the locally stored mask that tracks available context ids (which is currently also defined as `static uint64_t cid[OFI_CONTEXT_ID_SIZE]`).

Reply:
Can we figure out what this is actually being used for?  Is this conceptually a long bit array to record which peers have responded to a request?  I tried quickly looking at the code for this, but without delving deeper into the operation, I couldn't quickly figure it out (which is also a good indication that the existing code is unclear).

Reply:
We can defer this cleanup after merging the other changes.  The PR will need to be updated against the latest master, since the provider ops have changed.